src/manifest/container.rs

```diff
@@ -75,6 +79,238 @@ pub struct Metadata {
     pub version_id: u32,
 }
 
+wire_enum! {
+    /// A hash type for a manifest [`Toc`]
+    ///
+    /// Note that we currently only support the SHA-256 variant, even though
+    /// Cerberus permits SHA-384 and SHA-512 as well.
+    ///
+    /// [`Toc`]: struct.Toc.html
+    #[allow(missing_docs)]
+    pub enum HashType: u8 {
+      Sha256 = 0b000,
+      // Sha384 = 0b001,
+      // Sha512 = 0b010,
+    }
+}
+
+/// A TOC entry's raw bits.
+#[derive(Copy, Clone, PartialEq, Eq, Default, Debug, AsBytes, FromBytes)]
+#[repr(C)]
+struct RawTocEntry {
+    element_type: u8,
+    format_version: u8,
+    offset: u16,
+    len: u16,
+    parent_idx: u8,
+    hash_idx: u8,
+}
+
+/// An entry to a manifest's table of contents.
+///
+/// A TOC entry describes an element in manifest, such as its format and its
+/// location.
+///
+/// A TOC entry is encoded exactly the same way as this struct is laid
+/// out, byte-for-byte.
+///
+/// See [`Toc`](struct.Toc.html).
+pub struct TocEntry<'entry, 'toc, M> {
+    toc: &'entry Toc<'toc, M>,
+    // Invariant: index is always a valid index into `toc.entries`.
+    index: usize,
+    _ph: PhantomData<fn() -> M>,
+}
+
+impl<'entry, 'toc, M: Manifest> TocEntry<'entry, 'toc, M> {
+    #[inline]
+    fn raw(self) -> &'toc RawTocEntry {
+        &self.toc.entries[self.index]
+    }
+
+    /// Returns this entry's index in the TOC.
+    pub fn index(self) -> usize {
+        self.index
+    }
+
+    /// Returns the type of the element this entry refers to.
+    pub fn element_type(self) -> M::ElementType {
+        <M::ElementType as WireEnum>::from_wire_value(self.raw().element_type)
+            .expect("previously verified by check_invariants()")
+    }
+
+    /// Returns the format version of this TOC entry.
+    ///
+    /// Note that this is always a version Manticore supports;
+    /// Manticore will refuse to parse TOCs with versions it does not
+    /// support.
+    pub fn format_version(self) -> u8 {
+        self.raw().format_version
+    }
+
+    /// Returns a flash region indicating where this entry's data is located.
+    pub fn region(self) -> Region {
+        Region::new(self.raw().offset as u32, self.raw().len as u32)
+    }
+
+    /// Returns this entry's parent, if it has one.
+    pub fn parent(self) -> Option<Self> {
+        match self.raw().parent_idx {
+            0xff => None,
+            x => Some(
+                self.toc
+                    .entry(x as usize)
+                    .expect("previously verified by check_invariants()"),
```

Comment:
I feel like this is an `unreachable!()`, and the string is just bloating the binary. I don't think I care for now, though.

Reply:
I think the right thing to do is, if we find that such strings bloat the binary, add some kind of `unwrap!` macro where we can disable the error messages with a feature flag. Let's do that once we start running into size problems.